### PR TITLE
Add basic HTTP and ARM clients

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Respective clients and their constructors added to the azcore and arm packages.
Refactored some type definitions to avoid circular dependencies.

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

**Review by JeffreyRichter — APPROVED** [review]

**Review by chlowell — APPROVED** [review]

File: sdk/azcore/arm/client.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package arm
+
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	armrpolicy "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm/policy"
```
Comment (richardpark-msft) [inline sdk/azcore/arm/client.go:11]: What's the arm'r' here for?

**Review by richardpark-msft — APPROVED** [review]
Looks great, one minor question (might just be a typo)

File: sdk/azcore/arm/client.go (hunk 1 above)
Comment (richardpark-msft) [reply sdk/azcore/arm/client.go:11]: (ah, never mind, @chlowell  already found this one)

[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/arm/client.go]: Fixed

File: sdk/azcore/arm/client.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	ep := cloud.AzurePublic.Services[cloud.ResourceManager].Endpoint
+	if c, ok := options.Cloud.Services[cloud.ResourceManager]; ok {
+		ep = c.Endpoint
+	}
+	pl, err := armruntime.NewPipeline(pkg, moduleVersion, cred, runtime.PipelineOptions{
+		APIVersion: runtime.APIVersionOptions{
+			Location: runtime.APIVersionLocationQueryParam,
+			Name:     "api-version",
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/arm/client.go]: Removed redundant initialization.

File: sdk/azcore/arm/runtime/arm.go [hunk 3]
```diff
@@ -4,7 +4,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-package arm
+package runtime
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/arm/runtime/arm.go]: Renamed

File: sdk/azcore/internal/shared/shared_test.go [hunk 4]
```diff
@@ -140,3 +140,28 @@ func TestTransportFunc(t *testing.T) {
...
+
+	pkg, err = ExtractPackageName("malformed")
+	require.Error(t, err)
+	require.Empty(t, pkg)
+
+	pkg, err = ExtractPackageName(".malformed")
+	require.Error(t, err)
+	require.Empty(t, pkg)
+
+	pkg, err = ExtractPackageName("malformed.")
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/internal/shared/shared_test.go:164]: Added

File: sdk/azcore/internal/shared/shared.go [hunk 5]
```diff
@@ -133,3 +136,24 @@ type TransportFunc func(*http.Request) (*http.Response, error)
...
+	if !modVerRegx.MatchString(moduleVersion) {
+		return fmt.Errorf("malformed moduleVersion param value %s", moduleVersion)
+	}
+	return nil
+}
+
+// ExtractPackageName returns "package" from "package.Client".
+// If clientName is malformed, an error is returned.
+func ExtractPackageName(clientName string) (string, error) {
+	i := strings.IndexByte(clientName, '.')
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/internal/shared/shared.go]: Good catch, I didn't know about this one.

**Pull request merged**